src/api/ui/views/task.tsx

```diff
@@ -0,0 +1,227 @@
+import { rewriteTask, setTaskCompletion } from "ui/utils";
+import { Literal, Literals } from "expression/literal";
+import {
+    EditableAction,
+    EditableListField,
+    EditableState,
+    TextEditable,
+    useEditableDispatch,
+} from "ui/fields/editable";
+import { setInlineField } from "index/import/inline-field";
+import { BaseFieldProps } from "ui/fields/common-props";
+import { Field } from "expression/field";
+import { DateTime } from "luxon";
+import "styles/lists.css";
+
+export interface TaskProps extends ListState<MarkdownTaskItem | MarkdownListItem> {
+    /** task states to cycle through, if specified */
+    additionalStates?: string[];
+    /** fields to display under each item in this task list */
+    displayedFields?: (BaseFieldProps<Literal> & { key: string })[];
+}
+
+export function TaskList({
+    rows: items,
+    additionalStates: states,
+    renderer: listRenderer = (item, index) => <DefaultListElement element={item} />,
+    ...rest
+}: TaskProps) {
+    const content = useMemo(() => {
+        return (
+            <ul className="datacore contains-task-list">
+                {items?.map((item, ind) =>
+                    item instanceof MarkdownTaskItem ? (
+                        <Task state={{ ...rest, additionalStates: states }} item={item} />
+                    ) : (
+                        listRenderer(item, ind)
+                    )
+                )}
+            </ul>
+        );
+    }, [items, states]);
+    return <Fragment>{!!items && content}</Fragment>;
+}
+
+export function Task({ item, state: props }: { item: MarkdownTaskItem; state: TaskProps }) {
+    const app = useContext(APP_CONTEXT);
+    const core = useContext(DATACORE_CONTEXT);
+    const { settings } = core;
+    const states = [" ", ...(props.additionalStates || []), "x"];
+    const nextState = useMemo(() => {
+        if (props.additionalStates && props.additionalStates?.length > 0) {
+            let curIndex = states.findIndex((a) => a === item.$status);
+            curIndex++;
+            if (curIndex >= states.length) {
+                curIndex = 0;
+            }
+            return states[curIndex];
+        } else {
+            return item.$completed ? " " : "x";
+        }
+    }, [props.additionalStates, item, item.$status, item.$completed]);
+
+    const completedRef = useRef<Dispatch<EditableAction<Literal>>>(null);
+    const onChecked = useStableCallback(
+        async (evt: JSXInternal.TargetedMouseEvent<HTMLInputElement>) => {
+            // evt.stopPropagation();
+            const completed = evt.currentTarget.checked;
+
+            let newStatus: string;
+            if (evt.shiftKey) {
+                newStatus = nextState!;
+            } else {
+                newStatus = completed ? "x" : " ";
+            }
+            const parent = evt.currentTarget.parentElement;
+            parent?.setAttribute("data-task", newStatus);
```

Comment:
You should do this via react (by setting `data-task` to the newest status in your JSX), instead of directly mutating the DOM - React will not respect your DOM mutations so other re-renders will wipe this.